packages/foundation-ui/src/navigation/NavigationService.js

```diff
@@ -1,162 +1,58 @@
 import { EventEmitter } from "events";
 import { NAVIGATION_CHANGE } from "./EventTypes";
 
-class NavigationService extends EventEmitter {
-  constructor() {
-    super();
-
-    /**
-     * @typedef NavigationService~Context
-     */
-    const privateContext = {
-      instance: this,
-      deferredTasks: [],
-      definition: [
-        // The only default is root, so it goes first no matter what
-        { category: "root", children: [] }
-      ],
-      defer(method, args) {
-        this.deferredTasks.push({ method, args });
-      },
-      processDeferred() {
-        const tasks = this.deferredTasks.slice(0);
-        this.deferredTasks = [];
-
-        // If Task is unable to resolve at this point of time it will re-defer itself
-        tasks.forEach(({ method, args }) => {
-          method.apply(this.instance, args);
-        });
-      }
-    };
-
-    this.getDefinition = this.getDefinition.bind(privateContext);
-    this.registerCategory = this.registerCategory.bind(privateContext);
-    this.registerPrimary = this.registerPrimary.bind(privateContext);
-    this.registerSecondary = this.registerSecondary.bind(privateContext);
-
-    this.on(
-      NAVIGATION_CHANGE,
-      privateContext.processDeferred.bind(privateContext)
-    );
-  }
-
-  /**
-   * getDefinition - returns navigation definition
-   *
-   * @this {NavigationService~Context}
-   * @return {Array} definition items
-   */
-  getDefinition() {
-    return this.definition.slice(0);
-  }
-
-  /**
-   * registerCategory - register a Sidebar category
-   *
-   * @this {NavigationService~Context}
-   * @param  {String} category Category name
-   */
-  registerCategory(category) {
-    if (!this.definition.find(element => element.category === category)) {
-      this.definition.push({
-        category,
-        children: []
-      });
-
-      this.instance.emit(NAVIGATION_CHANGE);
+function unflatten(array, parent, tree) {
```

Comment:
This is never called with three arguments, so why is there a third arg?

Reply:
it's a random function from stackoverflow 😅 gotta optimize it

Reply:
Makes sense...I got the feeling wasn't written from scratch